File: ios/SpellIt/Engine/Crossword.swift
import Foundation

enum CrosswordDirection {
    case across
    case down
}

struct CrosswordPlacement: Identifiable {
    let word: String
    let hint: String
    var row: Int
    var col: Int
    let dir: CrosswordDirection
    var number = 0

    var id: String { "\(word)-\(row)-\(col)" }

    var cells: [GridCell] {
        (0..<word.count).map { i in
            GridCell(
                row: dir == .down ? row + i : row,
                col: dir == .across ? col + i : col,
            )
        }
    }
}

struct GridCell: Hashable {
    let row: Int
    let col: Int
}

struct CrosswordPuzzle {
    var placements: [CrosswordPlacement]
    var width: Int
    var height: Int
}

/// Port of the site's greedy crossword generator: perpendicular crossings
/// only, standard adjacency rules, compactness-scored retries.
enum CrosswordGenerator {
    static let maxWordLength = 9

    static func generate(pool: [WordEntry], target: Int = 5) -> CrosswordPuzzle {
        let usable = pool.filter { $0.word.count <= maxWordLength }
        var best: [CrosswordPlacement] = []
        var bestArea = Int.max
        let cozyArea = 120

        for _ in 0..<30 {
            let words = pickRandom(usable, min(usable.count, 12))
            guard let first = words.first else { break }
            var grid: [GridCell: Character] = [:]
            var dirs: [GridCell: Set<Bool>] = [:] // true = across
            var placed: [CrosswordPlacement] = []

            func setWord(_ entry: WordEntry, _ row: Int, _ col: Int, _ dir: CrosswordDirection) {
                let placement = CrosswordPlacement(word: entry.word, hint: entry.hint, row: row, col: col, dir: dir)
                let chars = Array(entry.word)
                for (i, cell) in placement.cells.enumerated() {
                    // Store lowercase so capitalized entries ("February") still
                    // cross lowercase words sharing the letter.
                    grid[cell] = Character(chars[i].lowercased())
                    dirs[cell, default: []].insert(dir == .across)
                }
                placed.append(placement)
            }

            func canPlace(_ word: [Character], _ row: Int, _ col: Int, _ dir: CrosswordDirection) -> Bool {
                let dr = dir == .down ? 1 : 0
                let dc = dir == .across ? 1 : 0
                if grid[GridCell(row: row - dr, col: col - dc)] != nil { return false }
                if grid[GridCell(row: row + dr * word.count, col: col + dc * word.count)] != nil { return false }
                var crossings = 0
                for i in word.indices {
                    let cell = GridCell(row: row + dr * i, col: col + dc * i)
                    if let existing = grid[cell] {
                        if existing != Character(word[i].lowercased()) { return false }
                        if dirs[cell]?.contains(dir == .across) == true { return false }
                        crossings += 1
                    } else {
                        if grid[GridCell(row: cell.row + dc, col: cell.col + dr)] != nil { return false }
                        if grid[GridCell(row: cell.row - dc, col: cell.col - dr)] != nil { return false }
                    }
                }
                return crossings > 0
            }

            setWord(first, 0, 0, .across)
            for entry in words.dropFirst() {
                if placed.count >= target { break }
                let wordChars = Array(entry.word)
                var options: [(Int, Int, CrosswordDirection)] = []
                for p in placed {
                    let pChars = Array(p.word)
                    for i in pChars.indices {
                        for j in wordChars.indices where pChars[i].lowercased() == wordChars[j].lowercased() {
                            let option: (Int, Int, CrosswordDirection) =
                                p.dir == .across
                                    ? (p.row - j, p.col + i, .down)
                                    : (p.row + i, p.col - j, .across)
                            if canPlace(wordChars, option.0, option.1, option.2) {
                                options.append(option)
                            }
                        }
                    }
                }
                if !options.isEmpty {
                    // Prefer compact placements.
                    func area(_ o: (Int, Int, CrosswordDirection)) -> Int {
                        var rows = placed.flatMap { [$0.row, $0.dir == .down ? $0.row + $0.word.count - 1 : $0.row] }
                        var cols = placed.flatMap { [$0.col, $0.dir == .across ? $0.col + $0.word.count - 1 : $0.col] }
                        rows += [o.0, o.2 == .down ? o.0 + wordChars.count - 1 : o.0]
                        cols += [o.1, o.2 == .across ? o.1 + wordChars.count - 1 : o.1]
                        return (rows.max()! - rows.min()! + 1) * (cols.max()! - cols.min()! + 1)
                    }
                    let sorted = options.sorted { area($0) < area($1) }
                    let pick = sorted.prefix(3).randomElement()!
                    setWord(entry, pick.0, pick.1, pick.2)
                }
            }

            let rows = placed.flatMap { [$0.row, $0.dir == .down ? $0.row + $0.word.count - 1 : $0.row] }
            let cols = placed.flatMap { [$0.col, $0.dir == .across ? $0.col + $0.word.count - 1 : $0.col] }
            let attemptArea = (rows.max()! - rows.min()! + 1) * (cols.max()! - cols.min()! + 1)
            if placed.count > best.count || (placed.count == best.count && attemptArea < bestArea) {
                best = placed
                bestArea = attemptArea
            }
            if best.count >= target, bestArea <= cozyArea { break }
        }

        // Normalize to 0-based coordinates and assign clue numbers.
        let minRow = best.map(\.row).min() ?? 0
        let minCol = best.map(\.col).min() ?? 0
        var normalized = best.map { p in
            var copy = p
            copy.row -= minRow
            copy.col -= minCol
            return copy
        }
        let height = (normalized.map { $0.dir == .down ? $0.row + $0.word.count - 1 : $0.row }.max() ?? 0) + 1
        let width = (normalized.map { $0.dir == .across ? $0.col + $0.word.count - 1 : $0.col }.max() ?? 0) + 1

        let starts = Set(normalized.map { GridCell(row: $0.row, col: $0.col) })
            .sorted { ($0.row, $0.col) < ($1.row, $1.col) }
        let numberByCell = Dictionary(
            uniqueKeysWithValues: starts.enumerated().map { ($0.element, $0.offset + 1) }
        )
        for i in normalized.indices {
            normalized[i].number = numberByCell[GridCell(row: normalized[i].row, col: normalized[i].col)] ?? 0
        }

        return CrosswordPuzzle(placements: normalized, width: width, height: height)
    }
}

File: android/app/src/main/java/com/skdaddle/spellit/engine/Crossword.kt
package com.skdaddle.spellit.engine

import com.skdaddle.spellit.model.WordEntry

enum class CrosswordDirection { ACROSS, DOWN }

data class GridCell(val row: Int, val col: Int)

data class CrosswordPlacement(
    val word: String,
    val hint: String,
    val row: Int,
    val col: Int,
    val dir: CrosswordDirection,
    val number: Int = 0,
) {
    val id: String get() = "$word-$row-$col"

    val cells: List<GridCell>
        get() = (0 until word.length).map { i ->
            GridCell(
                row = if (dir == CrosswordDirection.DOWN) row + i else row,
                col = if (dir == CrosswordDirection.ACROSS) col + i else col,
            )
        }
}

data class CrosswordPuzzle(
    val placements: List<CrosswordPlacement>,
    val width: Int,
    val height: Int,
)

/**
 * Port of the site's greedy crossword generator: perpendicular crossings
 * only, standard adjacency rules, compactness-scored retries.
 */
object CrosswordGenerator {
    const val MAX_WORD_LENGTH = 9

    fun generate(pool: List<WordEntry>, target: Int = 5): CrosswordPuzzle {
        val usable = pool.filter { it.word.length <= MAX_WORD_LENGTH }
        var best: List<CrosswordPlacement> = emptyList()
        var bestArea = Int.MAX_VALUE
        val cozyArea = 120

        for (attempt in 0 until 30) {
            val words = pickRandom(usable, minOf(usable.size, 12))
            val first = words.firstOrNull() ?: break
            val grid = mutableMapOf<GridCell, Char>()
            val dirs = mutableMapOf<GridCell, MutableSet<Boolean>>() // true = across
            val placed = mutableListOf<CrosswordPlacement>()

            fun setWord(entry: WordEntry, row: Int, col: Int, dir: CrosswordDirection) {
                val placement = CrosswordPlacement(
                    word = entry.word, hint = entry.hint ?: "", row = row, col = col, dir = dir,
                )
                val chars = entry.word.toCharArray()
                placement.cells.forEachIndexed { i, cell ->
                    // Store lowercase so capitalized entries ("February") still
                    // cross lowercase words sharing the letter.
                    grid[cell] = chars[i].lowercaseChar()
                    dirs.getOrPut(cell) { mutableSetOf() }.add(dir == CrosswordDirection.ACROSS)
                }
                placed.add(placement)
            }

            fun canPlace(word: CharArray, row: Int, col: Int, dir: CrosswordDirection): Boolean {
                val dr = if (dir == CrosswordDirection.DOWN) 1 else 0
                val dc = if (dir == CrosswordDirection.ACROSS) 1 else 0
                if (grid.containsKey(GridCell(row - dr, col - dc))) return false
                if (grid.containsKey(GridCell(row + dr * word.size, col + dc * word.size))) return false
                var crossings = 0
                for (i in word.indices) {
                    val cell = GridCell(row + dr * i, col + dc * i)
                    val existing = grid[cell]
                    if (existing != null) {
                        if (existing != word[i].lowercaseChar()) return false
                        if (dirs[cell]?.contains(dir == CrosswordDirection.ACROSS) == true) return false
                        crossings += 1
                    } else {
                        if (grid.containsKey(GridCell(cell.row + dc, cell.col + dr))) return false
                        if (grid.containsKey(GridCell(cell.row - dc, cell.col - dr))) return false
                    }
                }
                return crossings > 0
            }

            setWord(first, 0, 0, CrosswordDirection.ACROSS)
            for (entry in words.drop(1)) {
                if (placed.size >= target) break
                val wordChars = entry.word.toCharArray()
                val options = mutableListOf<Triple<Int, Int, CrosswordDirection>>()
                for (p in placed) {
                    val pChars = p.word.toCharArray()
                    for (i in pChars.indices) {
                        for (j in wordChars.indices) {
                            if (pChars[i].lowercaseChar() != wordChars[j].lowercaseChar()) continue
                            val option = if (p.dir == CrosswordDirection.ACROSS) {
                                Triple(p.row - j, p.col + i, CrosswordDirection.DOWN)
                            } else {
                                Triple(p.row + i, p.col - j, CrosswordDirection.ACROSS)
                            }
                            if (canPlace(wordChars, option.first, option.second, option.third)) {
                                options.add(option)
                            }
                        }
                    }
                }
                if (options.isNotEmpty()) {
                    // Prefer compact placements.
                    fun area(o: Triple<Int, Int, CrosswordDirection>): Int {
                        val rows = placed.flatMap {
                            listOf(it.row, if (it.dir == CrosswordDirection.DOWN) it.row + it.word.length - 1 else it.row)
                        } + listOf(o.first, if (o.third == CrosswordDirection.DOWN) o.first + wordChars.size - 1 else o.first)
                        val cols = placed.flatMap {
                            listOf(it.col, if (it.dir == CrosswordDirection.ACROSS) it.col + it.word.length - 1 else it.col)
                        } + listOf(o.second, if (o.third == CrosswordDirection.ACROSS) o.second + wordChars.size - 1 else o.second)
                        return (rows.max() - rows.min() + 1) * (cols.max() - cols.min() + 1)
                    }
                    val pick = options.sortedBy { area(it) }.take(3).random()
                    setWord(entry, pick.first, pick.second, pick.third)
                }
            }

            val rows = placed.flatMap {
                listOf(it.row, if (it.dir == CrosswordDirection.DOWN) it.row + it.word.length - 1 else it.row)
            }
            val cols = placed.flatMap {
                listOf(it.col, if (it.dir == CrosswordDirection.ACROSS) it.col + it.word.length - 1 else it.col)
            }
            val attemptArea = (rows.max() - rows.min() + 1) * (cols.max() - cols.min() + 1)
            if (placed.size > best.size || (placed.size == best.size && attemptArea < bestArea)) {
                best = placed.toList()
                bestArea = attemptArea
            }
            if (best.size >= target && bestArea <= cozyArea) break
        }

        // Normalize to 0-based coordinates and assign clue numbers.
        val minRow = best.minOfOrNull { it.row } ?: 0
        val minCol = best.minOfOrNull { it.col } ?: 0
        var normalized = best.map { it.copy(row = it.row - minRow, col = it.col - minCol) }
        val height = (normalized.maxOfOrNull {
            if (it.dir == CrosswordDirection.DOWN) it.row + it.word.length - 1 else it.row
        } ?: 0) + 1
        val width = (normalized.maxOfOrNull {
            if (it.dir == CrosswordDirection.ACROSS) it.col + it.word.length - 1 else it.col
        } ?: 0) + 1

        val starts = normalized.map { GridCell(it.row, it.col) }.toSet()
            .sortedWith(compareBy({ it.row }, { it.col }))
        val numberByCell = starts.withIndex().associate { (i, cell) -> cell to i + 1 }
        normalized = normalized.map {
            it.copy(number = numberByCell[GridCell(it.row, it.col)] ?: 0)
        }

        return CrosswordPuzzle(placements = normalized, width = width, height = height)
    }
}
